README.md
# zoomable_tilt_view
ZoomableTiltView is simple widget to add some juiciness to your app. 

![Example gif](gifs/example.gif)

## Usage

Simply import the package and wrap it inside ZoomableTiltView widget.

```dart
import 'package:zoomable_tilt_view/zoomable_tilt_view.dart';
```

## Example

```dart
ZoomableTiltView(
  yTiltLimit: 50,
  xTiltLimit: 50,
  tiltSensitivity: 1.5,
  zoomLimit: 1.8,
  child: Container(
    height: 200,
    width: 200,
    color: Colors.pinkAccent,
    child: Center(
      child: Text(
        "HIREME",
        style: TextStyle(
          fontSize: 24,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
      ),
    ),
  ),
),
```

## Parameters

| Parameter                           | Default                       | Description                                                                                       |
|-------------------------------------|-------------------------------|---------------------------------------------------------------------------------------------------|
| **yTiltLimit** int                      | 25                            | How much the child can tilt on Y axis\. Must be positive number\.                                 |
| **xTiltLimit** int                      | 25                            | How much the child can tilt on X axis\. Must be positive number\.                                 |
| **tiltSensitivity** double              | 1\.0                          | How much the child tilts\.                                                                        |
| **tiltResetAnimationDuration** Duration | Duration\(milliseconds: 900\) | How fast the tilt resets after you release the hold\.                                             |
| **zoomLimit** double                    | 1\.2                          | How much the child zooms when on press\.                                                          |
| **zoomAnimationDuration** Duration      | Duration\(milliseconds: 250\) | How long it takes to zoom in/out on press/release\.                                               |
| **zoomAlignment** Alignment             | Alignment\.center             | Which direction the child scales when pressed\. Useful if the child is near edges of the screen\. |


CHANGELOG.md
## [0.1.1] - 2020-01-23

* Added comments for better understandability


## [0.1.0] - 2020-01-02

* Added example to github


## [0.0.1] - 2019-12-24

* Releasing a reusable package from widget I made some time ago


test/zoomable_tilt_view_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:zoomable_tilt_view/zoomable_tilt_view.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/zoomable_tilt_view.dart
library zoomable_tilt_view;

import 'package:flutter/material.dart';

class ZoomableTiltView extends StatefulWidget {
  // How much view can be tilted on Y axis
  final int yTiltLimit;

  // How much view can be tilted on X axis
  final int xTiltLimit;

  // How easily the view tilts, default value 1.0
  final double tiltSensitivity;

  // How fast the tilt resets after you release the hold
  final Duration tiltResetAnimationDuration;

  // How much the child zooms when pressed
  final double zoomLimit;

  // How fast the child zooms in and out. Usually faster than tiltResetAnimationDuration
  final Duration zoomAnimationDuration;

  // Where the zoom starts, defaults to Alignment.center. If the view is on the right side of the screen you might want to use Alignment.right, for example.
  final Alignment zoomAlignment;

  final Widget child;

  ZoomableTiltView({
    Key key,
    @required this.child,
    this.yTiltLimit = 25,
    this.xTiltLimit = 25,
    this.tiltSensitivity = 1.0,
    this.tiltResetAnimationDuration = const Duration(milliseconds: 900),
    this.zoomLimit = 1.2,
    this.zoomAnimationDuration = const Duration(milliseconds: 250),
    this.zoomAlignment = Alignment.center,
  })  : assert(child != null),
        assert(yTiltLimit >= 0),
        assert(xTiltLimit >= 0),
        assert(tiltSensitivity >= 0),
        super(key: key);

  @override
  _ZoomableTiltViewState createState() => _ZoomableTiltViewState();
}

class _ZoomableTiltViewState extends State<ZoomableTiltView>
    with TickerProviderStateMixin {
  // Tilt animation
  AnimationController _xController;
  AnimationController _yController;

  // Zoom animation
  AnimationController _zoomController;
  Animation<double> _zoomAnimation;

  @override
  void initState() {
    super.initState();
    _zoomController = AnimationController(
      vsync: this,
      duration: widget.zoomAnimationDuration,
    );
    _zoomAnimation = Tween<double>(
      begin: 1.0,
      end: widget.zoomLimit,
    ).animate(CurvedAnimation(
      parent: _zoomController,
      curve: Curves.easeInCubic,
    ))
      ..addListener(() {
        setState(() {
          //
        });
      });

    _xController = AnimationController(
      vsync: this,
      lowerBound: -(widget.xTiltLimit).toDouble(),
      upperBound: widget.xTiltLimit.toDouble(),
      duration: widget.tiltResetAnimationDuration,
    );
    _xController.addListener(() {
      setState(() {
        //
      });
    });

    _yController = AnimationController(
      vsync: this,
      lowerBound: -(widget.yTiltLimit).toDouble(),
      upperBound: widget.yTiltLimit.toDouble(),
      duration: widget.tiltResetAnimationDuration,
    );
    _yController.addListener(() {
      setState(() {
        //
      });
    });

    _xController.value = 0.0;
    _yController.value = 0.0;
  }

  @override
  void dispose() {
    _zoomController.dispose();
    _xController.dispose();
    _yController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Transform.scale(
      alignment: widget.zoomAlignment,
      scale: _zoomAnimation.value,
      child: Transform(
        transform: Matrix4.identity()
          ..setEntry(3, 2, 0.001)
          ..rotateX(
              _yController.value * 0.01)
          ..rotateY(_xController.value * -0.01),
        alignment: FractionalOffset.center,
        child: GestureDetector(
          onPanUpdate: (details) {
            if ((_xController.value + details.delta.dx >=
                    -(widget.xTiltLimit)) &&
                (_xController.value + details.delta.dx <= widget.xTiltLimit)) {
              _xController.value = _xController.value +=
                  (details.delta.dx * widget.tiltSensitivity);
            }

            if ((_yController.value + details.delta.dy >=
                    -(widget.yTiltLimit)) &&
                (_yController.value + details.delta.dy <= widget.yTiltLimit)) {
              _yController.value = _yController.value +=
                  (details.delta.dy * widget.tiltSensitivity);
            }
          },
          onPanStart: (_) {
            _zoomController.forward();
          },
          onPanEnd: (details) {
            if (widget.yTiltLimit > 0) {
              _yController.animateTo(0.0, curve: Curves.easeOutBack);
            }
            if (widget.xTiltLimit > 0) {
              _xController.animateTo(0.0, curve: Curves.easeOutBack);
            }
            _zoomController.reverse();
          },
          child: widget.child,
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:zoomable_tilt_view/zoomable_tilt_view.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'ZoomableTiltView Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("zoomable_tilt_view"),
      ),
      body: Center(
        child: ZoomableTiltView(
          yTiltLimit: 50,
          xTiltLimit: 50,
          tiltSensitivity: 1.5,
          zoomLimit: 1.8,
          child: Container(
            height: 200,
            width: 200,
            child: Card(
              color: Colors.pinkAccent,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(24),
              ),
              elevation: 10,
              child: Center(
                child: Text(
                  "HIREME",
                  style: TextStyle(
                    fontSize: 24,
                    fontWeight: FontWeight.bold,
                    color: Colors.white,
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
